### PR TITLE
Allow CollisionObject2D to get shapes from tilemaps

### LINKED ISSUE #4454: RigidBody physics for TileMaps (CollisionShape passing to parent)
_feature proposal_
I was thinking about a kind of game, where you sort of build a ship - rooms, corridors, wings etc. The best way to do this would be with TileMaps. But there is an obvious problem.

To fly this ship you would have to treat it as a RigidBody, and there is no option like that. There is an option "Use Kinematic", but it's really confusing. Can I move() the TileMap ? How can I detect collisions with it?

The workaround right now would be to contain the TileMap in a RigidBody and for each cell create a collision shape suitable for a tile in that cell. The fact that it's doable this way, and not via a "TileMap" option is not a good thing in my opinion.

The whole idea of collisions within TileMaps seems a little flawed to me. My proposition:
1. Change "Use Kinematic" to "Use Static Body" selected by default
2. TileSets can contain collision shapes, but don't have to contain Static Body anymore
3. If the user deselects Use Static Body the TileMap will now have to be a child of a PhysicsBody2D (Rigid,Kinematic or Static) which will use the collision shapes

Since the tilemap can be very big I can understand why this was never a feature in the first place, but since an aformentioned workaround is possible... This would also make TileSet creation a little less of a pain (no need to use static bodies).
What do you think? Is this feature a good idea? Would it be hard to implement?

### LINKED ISSUE #22285: Area2D in tilemap
There is no way to add an area2D in the Tilemap. 
Immagine I want to do a platformer where there is water. All the water tiles are supposed to be an Area2D. 
It would be cool, if I can set up a specific tile as an area2d, So I can move and/or handle the objects enter that Area accordingly. 

What do you guys think about that?

### PR DESCRIPTION
**Pull request by bojidar-bg**
Fixes #4454 and likely resolves #22285.

There are a few additional cosmetic changes to CollisionShape2D. They can potentially be removed from the PR without impacting it.

The way this feature works is by adding a new option to TileMap: `collisions_use_parent` ("Use Parent"). If it is on, the Tilemap will put all its shapes into the parent CollisionShape2D body, which can be an Area2D, RigidBody2D, KinematicBody2D, or, tho that's useless, a StaticBody2D.

### CODE REVIEW

**Comment by bojidar-bg** [comment]
Rebased.

**Comment by akien-mga** [comment]
This is change is mostly limited to TileMap code, I think we don't need to wait for @reduz's approval (would have to if it required a major change of CollisionObject2D)..

So let's merge and test in the master branch :)

Since the PR is a bit old, do you want to rebase/retest it on current master to ensure that it still works as expected?

**Review by akien-mga — APPROVED** [review]

File: scene/2d/tile_map.cpp
Comment (akien-mga) [inline scene/2d/tile_map.cpp]: This is not covered by [our guidelines](http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes) so far, but I wonder if we should enforce that header includes should always use the "absolute" path from project root, e.g. `"scene/2d/area_2d.h"` to avoid confusion and have a better alphabetical ordering by path.

This would not apply to the .cpp file's own .h though.

Comment (bojidar-bg) [reply scene/2d/tile_map.cpp]: Fixed!

**Comment by bojidar-bg** [comment]
Rebased on master -- everything works just fine :smiley:.

**Comment by akien-mga** [comment]
Thanks!

**Comment by asheraryam** [comment]
If I understand this correctly, does it mean that in order to have different tilemap area detection (for example, separate detection for water and for spikes) then the way to accomplish it with this PR is using several separate tilemaps?

**Comment by bojidar-bg** [comment]
@asheraryam yes, that's correct. You might be able to handle multiple tiles with the same Area2D though, by checking local shape metadata.

**Comment by nightblade9** [comment]
This is great, thanks for working on this @bojidar-bg and pushing it through @akien-mga !

 I don't quite understand the explanation of how this works, so if this is not too much work, could I request:

1) A documentation change for this (assuming there wasn't one already)
2) Some sort of minimal MVP with, say, two tilemaps for spikes/water, that illustrates how it works

File: scene/2d/tile_map.cpp
Comment (akien-mga) [inline scene/2d/tile_map.cpp]: Was this removal intended? I don't see this bool being initialized anymore.

Strangely enough, running `doctool` now gives me:
```diff
diff --git a/doc/classes/TileMap.xml b/doc/classes/TileMap.xml
index ee5205efd..55666b94b 100644
--- a/doc/classes/TileMap.xml
+++ b/doc/classes/TileMap.xml
@@ -286,7 +286,7 @@
                <member name="collision_mask" type="int" setter="set_collision_mask" getter="get_collision_mask" default="1">
                        The collision mask(s) for all colliders in the TileMap.
                </member>
-               <member name="collision_use_kinematic" type="bool" setter="set_collision_use_kinematic" getter="get_collision_use_kinematic" default="true">
+               <member name="collision_use_kinematic" type="bool" setter="set_collision_use_kinematic" getter="get_collision_use_kinematic" default="false">
                        If [code]true[/code], TileMap collisions will be handled as a kinematic body. If [code]false[/code], collisions will be handled as static body.
                </member>
                <member name="collision_use_parent" type="bool" setter="set_collision_use_parent" getter="get_collision_use_parent" default="false">
```

Comment (bojidar-bg) [reply scene/2d/tile_map.cpp]: No idea, I don't remember touching it though.

Comment (akien-mga) [reply scene/2d/tile_map.cpp]: I fixed it in #30184.

**Comment by SuperRobinHood** [comment]
Hey, this is pretty cool! I've played around with it and it's really useful. In fact this feature is the only way I could have resumed my game development. A question: Is it possible to get the information about the actual overlapping tile? Say I have two trap tiles and I want both of them to have different effects. Can I detect which one the player walked on?

**Comment by bojidar-bg** [comment]
@SuperRobinHood Yes, it is possible, though a bit undocumented. You can get the "shape metadata" of the collision (mind it, not the shape resource's metadata). It should be a Vector2 containing the cell position of the shape.

Alternatively, you can have two separate collision shapes on each of the two tiles (with make unique), and use resource metadata, set a script on the shape resource, or even just compare the resource references directly.

**Comment by SuperRobinHood** [comment]
Looking forward to a deeper documentation then, because I can't manage to do it (I'm trying with Area2d against a StaticBody, but KinematicBody against Area2d would be useful as well)

**Comment by SuperRobinHood** [comment]
I think I've managed to do this	with:

var tile_position=collision.get_position()
var tilemap = collision.get_collider_shape()
var coordinates = tilemap.world_to_map(tile_position)
var tile=tilemap.get_cellv(coordinates)

However it seems to be returning tile -1 (that means nonexistent, right?) sometimes upon collision anyway. I think just creating a different tilemap might be a better way for now. Might have done it wrong though.

**Comment by SuperRobinHood** [comment]
I think I've managed to do this with:

```
var tile_position=collision.get_position()
var tilemap = collision.get_collider_shape()
var coordinates = tilemap.world_to_map(tile_position)
var tile=tilemap.get_cellv(coordinates)

However it seems to be returning tile -1 (that means nonexistent, right?) sometimes upon collision anyway. I think just creating a different tilemap might be a better way for now. Might have done it wrong though.

**Comment by SuperRobinHood** [comment]
I think I've managed to do this with:

```
var tile_position=collision.get_position()
var tilemap = collision.get_collider_shape()
var coordinates = tilemap.world_to_map(tile_position)
var tile=tilemap.get_cellv(coordinates)
```
It still seems to return tile -1 sometimes, though. I think I'm going to just create a different tilemap for now. But I might have done something wrong.

**Pull request merged**
